### PR TITLE
Add All Bell States task to Superposition kata

### PR DESCRIPTION
**Pull request by jkingdon-ms**
# Adding All Bell States Kata

This resolves part of https://github.com/microsoft/qsharp/issues/1185 .

### CODE REVIEW

File: katas/content/superposition_kata/all_bell_states/Verification.qs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    @EntryPoint()
+    operation CheckSolution() : Bool {
+        for index in 0 .. 3 {
+            if not CheckOperationsEquivalenceOnZeroStateAndIndexWithFeedback(
+                Kata.AllBellStates,
+                AllBellStates,
+                2,
+                index) {
+                Message("Incorrect");
```
Comment (tcNickolas) [inline katas/content/superposition_kata/all_bell_states/Verification.qs:29]: We can simplify this code using [partial application](https://learn.microsoft.com/en-us/azure/quantum/user-guide/language/expressions/closures#partial-application), and then we won't need the changes to the libraries at all!
```suggestion
            if not CheckOperationsEquivalenceOnZeroStateWithFeedback(
                Kata.AllBellStates(_, index),
                AllBellStates(_, index),
                2) {
```

We'll be using different signatures for the states we want to compare - sometimes integers, other times Booleans or arrays, so it's not really practical to introduce a new library for each signature of the operation we want to test. Instead, we bind the integer parameter of these operations to the index we want to test and leave qubit parameters unbound - the result is an operation with a single parameter that we can pass to a regular library. Partial application is very useful and especially in testing - let me know if you want more details, I'll try to dig up more docs.

(And with this, you can revert the change to the KatasLibrary.qs)

File: katas/content/superposition_kata/all_bell_states/Verification.qs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                Message($"The test case for index = {index} did not pass.");
+                return false;
+            }
+        }
+
+        Message("Correct");
```
Comment (tcNickolas) [inline katas/content/superposition_kata/all_bell_states/Verification.qs:35]: Let's cheer for the learner :-) (I like to keep this kind of messages uniform across all katas)
```suggestion
        Message("Correct!");
```

**Review by tcNickolas — CHANGES_REQUESTED** [review]
Looks good overall, I left a few comments. The most significant one regarding using partial application to fix index when testing instead of introducing new libraries - you'll be able to use this trick a couple of tasks later when dealing with preparing states with other additional classical variables.

Thank you!

**Comment by tcNickolas** [comment]
I've played with this a bit more, and my rewrite suggestion can be improved: right now if I just prep the first Bell state (index = 0), it prints Correct! before continuing to Incorrect on the next test. We can do it slightly different, print "Testing for index = 0..." before doing that code, so that the verdict Correct! is just for that test case, and then we don't need a separate failure message - Incorrect will say it for us. (You can see an example of that done in katas\content\random_numbers\random_n_bits\Verification.qs, verdict is printed by IsSufficientlyRandom)

File: katas/content/superposition_kata/all_bell_states/Verification.qs (hunk 1 above)
Comment (jkingdon-ms) [reply katas/content/superposition_kata/all_bell_states/Verification.qs:29]: good to know! thanks for the tip.

**Comment by jkingdon-ms** [comment]
> I've played with this a bit more, and my rewrite suggestion can be improved: right now if I just prep the first Bell state (index = 0), it prints Correct! before continuing to Incorrect on the next test. We can do it slightly different, print "Testing for index = 0..." before doing that code, so that the verdict Correct! is just for that test case, and then we don't need a separate failure message - Incorrect will say it for us. (You can see an example of that done in katas\content\random_numbers\random_n_bits\Verification.qs, verdict is printed by IsSufficientlyRandom)

Good point, I can update it like that for clarity.

File: katas/content/superposition_kata/all_bell_states/solution.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+$$\frac{1}{\sqrt2} \begin{bmatrix} 1 & 0 & 1 & 0 \\\ 0 & 1 & 0 & 1 \\\ 0 & 1 & 0 & -1 \\\ \underset{|\Phi^{+}\rangle}{\underbrace{1}} & \underset{|\Psi^{+}\rangle}{\underbrace{0}} & \underset{|\Phi^{-}\rangle}{\underbrace{-1}} & \underset{|\Psi^{-}\rangle}{\underbrace{0}} \end{bmatrix}$$
+
+
+Notice that each of the columns in the unitary matrix corresponds to one of the Bell states.
+This unitary transformation transforms the computational basis into the Bell basis, which is exactly what the task asks us to do.
+
+We see that this transformation converts $|00\rangle$ into the first Bell state, $|01\rangle$ into the second Bell state, etc. 
+We just need to make sure we set the qubits to the correct state before applying this transformation, using $X$ gates to change the initial $|0\rangle$ states to $|1\rangle$ if needed. 
+
+In Q#, we can use the <a href="https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.convert.intasboolarray">IntAsBoolArray</a> function to convert the input `index` to the right bit pattern.
```
Comment (tcNickolas) [inline katas/content/superposition_kata/all_bell_states/solution.md:60]: The API docs format changed a bit with the Modern QDK, sorry I forgot to check that on the first round of review!
```suggestion
In Q#, we can use the <a href="https://learn.microsoft.com/qsharp/api/qsharp-lang/microsoft.quantum.convert/intasboolarray">IntAsBoolArray</a> function to convert the input `index` to the right bit pattern.
```

**Review by tcNickolas — APPROVED** [review]
Looks great! Just one link discrepancy I missed on the first review - I think I can resolve it myself by applying a suggestion.

Thank you!

**Pull request merged**